### PR TITLE
NativeTokenDestination burn is a simple transfer

### PR DESCRIPTION
**Pull request by geoff-vball**
## Why this should be merged
Removes the separate burn mechanism for the native token bridge. This will be closer to the unified interface for moving these to the `teleporter-token-bridge` repo, and does most of the heavy lifting before implementing the reward mechanism for reporting burned tx fess.

## How this works
NativeTokenSource and ERC20TokenSource no longer receive separate burn messages.

NativeTokenDestination will take the difference between the last time it reported the amount of burned fees, and send an unlock message to send those fees to a burn address.

## How this was tested
Existing tests

### CODE REVIEW

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (minghinmatthewlam) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:217]: why did we remove Teleporter fees here?

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (minghinmatthewlam) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:220]: don't we want this to go to the blackhole address so that the sources don't need specific handling of burn report transfers?

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (minghinmatthewlam) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:46]: nit: I prefer the old name, `TRANSFER_TO_SOURCE_BURN_ADDRESS` or similar. Also, can the comment above be updated.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (geoff-vball) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:220]: `subnet-evm` actually disables transfers the the blackhole address. So while we could send ERC20s to that address, we couldn't send native tokens there.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (geoff-vball) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:46]: We use it on both chains. On the destination chain as well to burn native tokens before transferring to the source chain.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/INativeTokenDestination.sol
Comment (michaelkaplan13) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/INativeTokenDestination.sol:43]: The dev comment here needs updating now.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (michaelkaplan13) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:44]:
```suggestion
     * @notice Address used by this contract to blackhole funds, effectively burning them.
     * @dev Native tokens are burned by this contract by sending them to this address when transferring tokens back to
     * the source chain. When reporting burned transaction fee amounts, this address is used as the recipient
     * address for the funds to be sent to be burned on the source chain.
     * Different from BURNED_TX_FEES_ADDRESS so that the total amount burned in transaction fees and the amount 
     * total amount burned to be sent back to the source chain can be tracked separately. 
```

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (michaelkaplan13) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:208]: Comment for future consideration in a later PR:

I think we should consider having `allowedRelayerAddresses` be something that is set in Teleporter application constructors, with the managers of the application being allowed to update it. IMO, the use case for `allowedRelayers` is if an application wants to restrict who is able to relay it's messages to provide off-chain checks (on validity, compliance, etc). I don't think it makes sense for message senders (i.e. end users calling the applications) to be able to set it themselves, since they only care about anyone delivering their message, but not who.

Comment (michaelkaplan13) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:208]: cc @minghinmatthewlam for you to think about also as we move forward with the Teleporter bridge contract implementations.

Comment (geoff-vball) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:208]: Yeah I do like the idea of abstracting it away from the users.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (cam-schultz) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:50]: I think it would be clearer if they were two separate variables, even if they contained the same address (I don't have a strong preference on if they should or not)

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (cam-schultz) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol]: `Used when reporting burned gas fees` seems more appropriate for `BURNED_TX_FEES_ADDRESS`

Are fees handled by this address? My understanding is that it's used to
1) burn tokens on the destination to be unlocked on the source
2) burn tokens on the source when reporting burned fees

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (cam-schultz) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:217]: Can we leave the Teleporter fees in here then, and remove them in the follow up PR?

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/TokenSource.sol
Comment (cam-schultz) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/TokenSource.sol:74]:
```suggestion
```

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/TokenSource.sol
Comment (cam-schultz) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/TokenSource.sol:76]: We may want to rename this function to something more general since it now handles unlocks and burns (even though burning is implicit by sending to the burn address). Perhaps `_transferTokens`

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (minghinmatthewlam) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:212]: would this make sense in `TeleporterUpgradeable`? Or do we want to leave it up to the dApps to implement and not set as a default

Comment (minghinmatthewlam) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:212]: hmm what happens for a contract that gets called by other dApps, who have their individual allowed relayer addresses though?

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/TokenSource.sol
Comment (geoff-vball) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/TokenSource.sol:76]: I think `unlock` is the most general. We've locked the tokens, and are being told to unlock them to an address by the destination. We don't really care if that chain is "burning" them or not

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (michaelkaplan13) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol]: > would this make sense in TeleporterUpgradeable? Or do we want to leave it up to the dApps to implement and not set as a default

I think it likely makes sense as a separate utility. Along the lines of a `RestrictedRelayerSender`.

> hmm what happens for a contract that gets called by other dApps, who have their individual allowed relayer addresses though?

Sorry not sure I follow. The dApp calling `sendCrossChainMessage` is what gets to define the allowed relayers for its own messages since they may have their own restrictions.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (michaelkaplan13) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:233]: We need to increment `latestBurnedFeesReported` here. Lets also add a unit test that would catch this case by calling `reportBurnedTxFees` multiple times. 

I think we should require that the `scaledAmount` is non-zero also.

**Review by cam-schultz — COMMENTED** [review]
LGTM once the open comments are addressed

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (minghinmatthewlam) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol]: Lets say we have contract A that gets called by a bunch of other contracts B, C, D, E...If contract A constructor takes in empty allowed relayers list, so any relayers allowed, but the other contracts (B,C,D...) have their own `allowedRelayerAddresses` list they want to pass down to the call of contract A.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (cam-schultz) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:235]: Should this be `latestBurnedFeesReported=burnedDifference`?

Comment (geoff-vball) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:235]: No, each call of this function should report the incremental change since the last time it was called, so we save the lastest `burnedAddressBalance`. I can rename the variable to `latestBurnedAddressBalance` if that would be more clear.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (cam-schultz) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:234]: I think we should `require` that `burnedAddressBalance >= latestBurnedFeesReported`, or use safemath. If the destination somehow becomes corrupted such that `BURNED_TX_FEES_ADDRESS.balance` decreases, that would cause an underflow and, and up to INTMAX reported burned fees to the source chain.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (geoff-vball) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:235]: Yeah, renamed.

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (geoff-vball) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:234]: Yeah, I don't think this should be possible under any circumstance, but having the sanity check is worth it based on the catastrophic consequences.

Comment (cam-schultz) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol]: Thinking about it some more, a scenario in which the destination chain is corrupted such that the burn address is compromised, but arbitrary Teleporter messages are still not able to be sent seems not that likely. Still a good idea to prevent an invalid value here. Thanks for the fix.

**Review by cam-schultz — APPROVED** [review]

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (minghinmatthewlam) [inline contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:60]: do we need this duplicate addrses variable? Can we just use `BURN_FOR_TRANSFER_ADDRESS` in reporting burned fees

Comment (geoff-vball) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol:60]: https://github.com/ava-labs/teleporter/pull/347#discussion_r1523699992

**Review by minghinmatthewlam — APPROVED** [review]

File: contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol
Comment (minghinmatthewlam) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol]: Reading over again, I think i'm in agreement for dApps that want to restrict relayers, they can set it in their constructor. For dApps that don't care about restricted relayers, they can still take in `allowedRelayerAddresses` as a parameter in sends.

Comment (michaelkaplan13) [reply contracts/src/CrossChainApplications/examples/NativeTokenBridge/NativeTokenDestination.sol]: Yeah I think it makes sense for the immediate caller of `sendCrossChainMessage` (i.e. the cross-chain dApp itself) to limit the allowed relayers if necessary. Exposing that option to callers of the dApp makes it possible for users of the cross-chain dApp to limit relayers for their own messages, but that's not something I would generally recommend for dApps, since users should not need to know of or about relayers at all.

**Pull request merged**
